### ISSUE TITLE
👉 Question: What are the main pain points you have using Strapi? (Episode 4)

### ISSUE THREAD

**Issue by derrickmehaffy** [strapi/strapi#13430]
Three years ago the Strapi Chief Product Officer @Aurelsicoko asked the community what their [main pain points](https://github.com/strapi/strapi/issues/3415) were and how critical these issues were to the rest of the community.

Since then, every year ([2021](https://github.com/strapi/strapi/issues/10419), [2020](https://github.com/strapi/strapi/issues/6634), [2019](https://github.com/strapi/strapi/issues/3415)) for the month of June we have asked the same question.

This year is a special one given the release of Strapi v4 and the process of deprecating Strapi v3. Once again the Strapi team wish to get the broad community feedback, so the question again this year is: **what are the main pain points you have using Strapi?**

Some Rules for this issue: 
- Please try to be short and to the point and describe your use-case and why you feel this is an important issue. 
- Only one pain point per comment (you can add as many comments as you want, but only one per comment please)
- Be sure to construct pain points respectfully and constructively, following our [Code of Conduct](https://forum.strapi.io/faq)

-----

### Instructions

- Please only post one pain point at a time.
- If your pain point has already been listed, make sure to **upvote it** instead of posting a similar message. A simple :+1:  will suffice.

### Template

- **Issue:** Please describe your issue.
- **Context:** Give more details about the use case.

Copy paste the following markdown template:
```markdown
### Pain point

**Issue:** 
Please describe your issue.

**Context:** 
Give more details about the use case.

**Which version are you on:**
v3 or v4
```

-----

We look forward to reading all your comments. All contributions are welcomed, don't be shy 🙈

**Comment by jameshhood** [comment]
### Pain point

**Issue:** 
Planning - Previewing - Publishing Groups of changes

**Context:** 
I have recently swapped from using prismic to strapi and we use the CMS tool in our e-commerce site. The challenge I run into is the ease of being able to publish something and still being able to have a "work-in-progress" version of the same document before actual publishing it. If I have a group of changes I want to all go live at once, there is also no way to group all the changes into a feature release/version and deploy them all at once. 

It would be fantastic to be able to work on a live preview and then either schedule by a date/time or manually publish a group of changes instead of just working on each document at a time and just publish/unpublish.

Prismic does this with labels/releases just like github and I guess they have a server scheduler that can auto deploy a particular set of changes in said labels/releases

**Which version are you on:**
v4

**Comment by jameshhood** [comment]
### Pain point

**Issue:** 
Need a color picker UI component

**Context:** 
I store a lot of theme configuration for my eCommerce site so we can change those values through CMS but there is no good way to visually see what that color is in the editor or be able to select one. A color picker component would be great to have here to return colors in hex

**Which version are you on:**
v4

**Comment by ipseidev** [comment]
### Pain point

**Issue:** 
Need to be able to sort relations and media

**Context:** 
I NEED, to have the possibility to order my media or my relations (articles on a feed for example). This is the biggest pain point for me ATM

**Which version are you on:**
both

**Comment by dogunbound** [comment]
### Pain point

**Issue:** 
Similarly named components in Dynamic zones disappearing. 

**Context:** 
[Give more details about the use case.](https://www.youtube.com/watch?v=j-TYly0s0L0&t=54s)

https://github.com/strapi/strapi/issues/11955

**Which version are you on:**
v3 and v4

On v3, this caused it to crash. It improved in v4 because it silently failed instead of quietly failing.

**Comment by wintercounter** [comment]
All my projects using Strapi still stuck on v3, so I cannot add anything on v4.

There are 3 main reasons I left Strapi for a competitor:

1. Too late migration guide, still missing migration scripts.
2. Awful development experience with the reload times. On a small size project with Docker on Windows (no WSL) startup time is around 40-60s. With WSL 20-30s. On native Linux it's 10-15s. Still unacceptable. (Remote database).
3. Performance - I know the data-layer has been rewritten in v4, but from what I've seen I'm pretty sure there are no huge perf improvements. Maybe I could query components also now?!

All in all, Strapi was a huge disappointment for me the last 2 years and today I pretty much regret my choice. It became the worst part of all the projects I used for. Sad but true.

**Comment by jurjendevries** [comment]
Pain point:
Great initiative I found out by your discord message. For me the interface feels a bit slow. The part of the idea of headless is blazing fast websites. The website is fast, but for editing content it should be fast too. For example after login it shows me a loader for 5 till 10 seconds. It would be better to have no loader at all, and have it in milliseconds.

Context:
CMS interface

Which version are you on:
v4.1.12 - Node version v14.19.3

**Comment by jurjendevries** [comment]
Pain point:
The updating process to a newer version feels too technical and it makes me feel insecure. I have to edit the version at 4 different places in one package.json file. The insecure part is that I don't know what to do with plugins such as @strapi/plugin-graphql . Should I also update it too the newer version, and better-sqlite3  when should I update that version, when is it outdated and so.

Context:
updating process

Which version are you on:
v4.1.12 - Node version v14.19.3

**Comment by Redemption198** [comment]
### Pain point

**Issue:** 
Use Strapi on pre existing databases

**Context:** 
It would be nice if Strapi could "automagically" wrap around existing databases with real-time changes to the structure made from Strapi or by directly modifying the database.

**Which version are you on:**
both

**Comment by LouisLec** [comment]
### Pain point

**Issue:** 

Admin panel is sooo slow when you have multiple relations in your entities, like in here https://forum.strapi.io/t/slow-content-manager-in-bigger-project/16062

Also: I would like to have key-values for enums, so the admin panel users can select a value written in plain english and not a no_blank_space_no_special_character one.

Also: documentation for Lifecycle hooks is too poor

**Context:** 

I have an app with many entities related to each other and query time in the admin grows bigger as I add items to my collections. The GraphQL requests are 100% OK, but loading an item in a collection to modify it can take up to 20sec which is too long. We reduced that loading time by reworking some of our relations (make it unidirectional) which improves a bit loading time but is detrimental to the user experience as the relation can be updated/created only from one of the two collections. Please help 👏🏼 

Additionnal context: we moved from v3 to v4 as the API was too slow. API is way faster in v4 but admin panel was faster in v3

**Which version are you on:**
v4

**Comment by nickvanballegooijen** [comment]
**Issue:** 
The [guides for unit testing](https://docs.strapi.io/developer-docs/latest/guides/unit-testing.html#install-test-tools) are outdated. I can hardly understand that Strapi is updated to v4 but the documentation is not up-to-date. How are we supposed to develop without documentation?

**Context:** 
U want to write unit tests for sparing time with testing my code.

**Which version are you on:**
v4

**Comment by fschucht** [comment]
### Pain point

**Issue:** 
Working with Strapi on smaller screens

**Context:** 
I often edit content from my iPad and the navigation already takes about ~1/3 of the screen while reducing the usable space for editing the content and shifting it off center. It would be great if the navigation could be collapsed so that the main focus is editing the content.

**Which version are you on:**
v4

**Comment by fschucht** [comment]
### Pain point

**Issue:** 
Managing dynamic zones with lots of components can be quite cumbersome

**Context:** 
I use dynamic zones quite extensible to store content for blog posts on my website. A typical blog post on my site has about 50-100 components in a dynamic zone. I don't use markdown as I have lots of custom components, like linked posts and info boxes. Also, markdown doesn't support some use cases, like checklists with custom bullet styles.

However, adding and updating content is quite cumbersome for multiple reasons:

1. It is not visible what content the component holds as all of them have the same name. Therefore, finding the right component requires me to expand each of them until I find the correct one.
2. If I want to add a component below another one I have to add one to the end of the dynamic zone and then manually move it using the arrows of the component.

Ideally, each element would have some sort of content preview to fix the first issue.

Regarding the second issue I see two solutions:
- Ideally, there would be drag & drop support for moving elements around. It could work the same way as it already does for lists.
- As an alternative, there could be a "Add new component before or after this one" button on each element.

**Which version are you on:**
v4

**Comment by finnsnape** [comment]
> ### Pain point
> 
> **Issue:** Working with Strapi on smaller screens
> 
> **Context:** I often edit content from my iPad and the navigation already takes about ~1/3 of the screen while reducing the usable space for editing the content and shifting it off center. It would be great if the navigation could be collapsed so that the main focus is editing the content.
> 
> **Which version are you on:** v4

I'd like to add that this is the reason I can't use Strapi at all. In my view, it's unreasonable to expect non-technical clients to use an interface that is not responsive in `{{ currentYear }}`.

**Comment by Cpaczek** [comment]
### Pain point

**Issue:** 
Strapi Update with Plugins

**Context:** 
I am pretty sure this is in progress but the marketplace needs to be integrated within Strapi. With this plugins should be verified to work with a specific version to prevent incompatibility when upgrading Strapi.

I think plugins should be verified with a certain version in 3 ways

-  Community Votes: This allows the community to "vote" if a plugin is still compatible which maybe faster than the author confirming it
- Author confirmation: the plugin author confirms the plugin is compatible
- Automated testing: Strapi should be able to run automated tests (or at least have the plugin author do it with a webhook with github actions or smth)

All 3 of these options should be displayed on the plugin page and once the `update` cli command is released it should give warnings about incompatible plugins.

**Which version are you on:**
v3 or v4

**Comment by tilman** [comment]
## Pain point
**Issue:**
No nesting of Dynamic Zones is possible

**Context:**
Nesting of Dynamic Zones (at least 2 levels) would greatly enhance the flexibility in which we can create layouts

Which version are you on:
v4

**Comment by tilman** [comment]
## Pain point
**Issue:**
data transfer/sync between environments (preview, dev, staging, testing)

**Context:**
A modern developer expierence includes deploying preview branches for each feature (like vercel does). For every feature branch we have to deploy strapi. Right now there is no good way how to transfer content from one deployment to another.

Which version are you on:
v4

**Comment by msoler75** [comment]
Context: backend programmatically customization

Issue: the syntax of three API -service, entityservice and query engine- are very similar but with bit differences that confuse and delay the learning curve. You have to learn and understand 3 so almost identically syntax but with particularities. 

That make me the programming very slow until I got the exact idea and I filled in my knowledge gaps. The lack of chart of syntax comparison examples in documentation make it more difficult.

When you begin to learn a full API documentation of a CMS like strapi, you have to learn many things. I think it is desirable that strapi will be very intuitive to program in backend too. 

I suggest to universalize syntax or allow some keywords be accepted in all three API. e.g. populate: 'true' or populate '*' allowed in all API. Allow 'sort' in query engine API, and so on.

Explanation & detail:
(not tested so can have errors)

```
// FINDONE

// SERVICE
const article = await strapi.service('api::article.article').findOne(1, // id field
    {
        populate: {
            author: '*', // 'asterisk' mark to populate all
            populate: {
                image: {
                    fields: ['url', 'width', 'height'] // 'fields' to select fields
                }
            }
        }
    })
// returns one object with the content or null

// ENTITYSERVICE
const article = await strapi.entityService.findOne('api::article.article', 1, // id field
    {
        populate: {
            author: true, // 'true' mark to populate all
            populate: {
                image: {
                    fields: ['url', 'width', 'height'] // 'fields' to select fields
                }
            }
        }
    });

// QUERY ENGINE
const article = await strapi.db.query("api::article.article").findOne({ // no id field, but where clause
    where: { // to get id
        id: {
            $eq: 1
        }
    },
    populate: {
        author: true, // 'true' mark to populate all
        populate: {
            image: {
                select: ['url', 'width', 'height'] // 'select' to select fields
            }
        }
    }
})
// returns one object with the content or null


// FIND / FINDMANY

// SERVICE
const {
    results: articles
} = await strapi.service('api::article.article').find({
    // 'filters' to filter
    filters: {
        category: { $gt: '2021-11-17T14:28:25.843Z' },
    },
    // 'sort' field to sort
    sort: {
        createdAt: 'DESC'
    }
})
// returns {results, meta} array


// ENTITYSERVICE
const articles = await strapi.entityService.findMany('api::article.article', {
    // 'filters' to filter
    filters: {
        createdAt: { $gt: '2021-11-17T14:28:25.843Z' },
    },
    // 'sort' field to sort
    sort: {
        createdAt: 'DESC'
    }
})
// returns array


// QUERY ENGINE
const articles = await strapi.db.query('api::article.article').findMany({
    // 'where' to filter
    where: {
        createdAt: { $gt: '2021-11-17T14:28:25.843Z' },
    },
    // 'orderBy' field to sort
    orderBy: {
        createdAt: 'DESC'
    },
  })
// returns array
```

#FindOne#
|                     |      Service    |  EntityService  |  Query Engine  |
|--------------|:-------------:|----------------:|----------------:|
| id                |        id field   |     id field        |   where  param   |
 | populate all |        *           |       true          |       true          |
| select fields |        fields    |       fields         |       select          |
| return         |  single object |  single object |  single object  |


#Find / FindMany#
|                     |      REST API   | Service    |  EntityService  |  Query Engine  | 
|--------------|:-------------:|----------------:|----------------:|----------------:|
| filters           |      filters      |              |      filters       |     filters          |   where            |        
 | sort results |      sort       |         sort       |       sort          |       orderby       |
| select fields |      fields     |         fields    |       fields         |       select          |
| return         | {data}         |  {results} |  array            |  array                |

**Comment by p741633** [comment]
Pain point

Issue:
Cannot create folders in Media Library

Context:
Organized alot of Media Library in folders would be the most appropiate

Which version are you on:
v4

**Comment by fadhilx** [comment]
### Pain point

**Issue:** 
Unresponsive UI, bad design system to customize

**Context:** 
The UI is good, but only for desktop, switch to mobile view, and it will be unusable at all, i dont expect strapi to be able to build a schema on mobile, but at least allow non-tech to edit the content.

**Which version are you on:**
v4

**Comment by dmurphy779** [comment]
### Pain point

**Issue:** 
Mobile layout for editors

**Context:** 
It's been said before and I think it's on the docket, but +1 for mobile layout for editor view. I know having a separate front end is the suggested standard but many times I just need a simple app for the client. 

**Which version are you on:**
v4

**Comment by simongroenewolt** [comment]
### Pain point

**Issue:** 
Selecting a relation is hard if the list of options is long. I'd like to have a dedicated 'browser' to select relations when this list is above a certain length. This could function and look similar to the media library, with (customizable) search options to locate the content you want to link.

**Context:** 
We are using relations where you might have to choose from hundreds or thousands of options. A select-box doesn't really work for that. Right now I'm doing most of the linking in code, but it would be really convenient to be able to use the admin ui as well.

**Which version are you on:**
v4

**Comment by simongroenewolt** [comment]
### Pain point

**Issue:** 
Allow relations to be shown in the 'main content area' instead of in the sidebar only. (And maybe, while you're at it, allow normal fields in the sidebar?)

**Context:** 
In our case some relations of an object are just as important as its 'normal' fields, we'd like to reflect that by showing them in the center. Also: if you have a lot of them the sidebar becomes cramped.

**Which version are you on:**
v4

**Comment by paulpopus** [comment]
### Pain point

**Issue:** 
We need more developer control over the admin interface. The ability to conditionally require, fill or hide fields. This includes ability to move the relational fields into the main content area and have other things like fieldsets and tabs available on the admin interface for grouping fields and conditional events.

**Context:** 
When building complex interfaces or simply trying to streamline data entry for a variety of users, we often have interdependent data. Currently our best solution is patch-package and overriding the base modules. For us this is where platforms like Sanity.io really shine in ease of use and implementation.

**Which version are you on:**
v4

**Comment by paulpopus** [comment]
### Pain point

**Issue:** 
Make media content type as a customisable as any other entity.

**Context:** 
Media items such as images or videos managed in the Media library are currently not customisable with the content type builder (or any other way) in the case that we want to enrich those content items with other fields like categories and so on that one would maybe be allowed to the filter them by in the media library.

**Which version are you on:**
v4

**Comment by paulpopus** [comment]
### Pain point

**Issue:** 
Better configuration specific to field types and allowing all fields to be required in addition to rules for multi selects like min and max rules.

**Context:** 
Currently it feels like certain field types like the Time field, lack important configuration options like Step amount (the only way is patch-package) or for the default value to be the "time now" or "date today".

**Which version are you on:**
v4

**Comment by PatrikTheDev** [comment]
### Pain point

**Issue:** 
users-permissions plugin isn’t easily configurable from any config files, I want to define things to allow while unauthenticated before the actual deployment. This should be exposed somewhere (or at least allow me to set find and findOne to work for unauthenticated by default).

**Context:** 
I’m building a CMS system with Next.js and Strapi

**Which version are you on:*
v4

**Comment by Lullu08** [comment]
### Pain Point
**Issue:**
Text Editor is markdown first and not configurable

**Context:**
I know that you can switch out the default editor, however using ckeditor (which is way richer than what we need) really slowed down performance. On big collection types with huge dynamic zones it was (at least when not collapsed) not even possible anymore to make changes. Finding the perfect editor is very cumbersome and if there is one that seems appropriate for our use case it is mostly not compatible with Strapi. It would be very helpful if one could decide whether to use markdown or richtext as default and configure options shown in the editors toolbar.

**Which version are you on:**
v3 and v4

**Comment by Lullu08** [comment]
### Pain Point

**Issue:**
Populating Enumeration Field with external Data

**Context:**
This will probably be more easily done when customizable fields are introduced, however I can imagine that providing a selection field that can be populated by an external API or imported values is something a lot of people could make use of to use Strapi in their respective contexts and should be an existing field type. 

**Which version are you on:**
v3 and v4

**Comment by DrMint** [comment]
### Pain point

**Issue:** 
Being able to customize the "Responsive friendly upload" function (this function generates multiple formats of the uploaded images):
- Being able to create multiple presets each with the following options:
  - Preset name (used to retrieved this version of the image)
  - Max width
  - Max height
  - Quality
  - Format (at least JPG and WebP)

**Context:** 
I've being using Directus in the past, they have implement a similar system (see https://docs.directus.io/configuration/project-settings/#files-storage). Since, I've switch to Strapi but found the current "Responsive friendly upload" lacking. Thus I disabled it and I've created an external tool using the Strapi media webhooks to trigger my tool whenever an image is upload/edited/deleted. But it would be sweet to have this feature directly in Strapi.

**Which version are you on:**
v4

**Comment by paulpopus** [comment]
Pain point

Issue:
Admin UI menu customisation, maybe even per role. Similar to Sanity.io's admin UI, or Drupal's views.

Context:
First usability issue is that the current menu structure isn't "content first", in comparison with a platform like Sanity.io, where we can customise the menu to create a custom context for our editors. 

For example, sometimes you want to split up your content by certain queries, in our use case we have different types of events and ideally we can create listings such as "Upcoming Events" and "Past Events" and not just have the "Events" menu item we currently have. Sanity does it by allowing you to list items using queries, Drupal does this by having custom "views" for listings.

Similarly, we should be able to hide parts of the UI for certain roles. If you're a content editor you absolutely do not need to see anything else in Strapi other than content, so having the "content manager" menu first is redundant.

Another example is if the menu were instead customisable with queries we could create an "Events Hub" menu item, which would link to a single type and then as children we'd have various views by event type and a link to "Create event" directly.

Which version are you on:
v4

**Comment by Godwin004code** [comment]
I've been trying to fix the stale request error , but I've not seen a perfect answer online .

**Comment by rktyt** [comment]
### Pain point

**Issue:** 
Can't use ESM.

**Context:** 
The admin panel uses webpack, but the server-side code does not, so you have to prepare two files to share the code, which is a hassle.

**Which version are you on:**
v4

**Comment by Sandstedt** [comment]
### Pain point

**Issue:**
The localisation plug-in is not suitable for larger organisations

**Context:**
Working with localisation firms, inviting the translator to the admin is not at all a good solution. They need a file that they can import to their translation software, where they keep market and client specific dictionaries to quickly translate common terms and phrases. And also make the language consistent overtime (so ex a certain custom hardware tool is named the same among all pages for that language).

They will also need to be able to see what has updated, and in what context it exist, ex a PDF of the page it exist on.

Other missing features in the localisation plugin is the ability to sync changes from the default language. Ex if you add a new component, or update an image, you need to be able to sync that change (with a manual review) to all other localisations.

It would also be beneficial to be able to select multiple languages for a page. Ex a page in Spanish could have both Spanish (Chile) (es-CL) and Spanish (Spain) (es-ES). Sometimes the content could differ due to being different markets, but most of content can probably be shared among the two different locales.

**Which version are you on:**
v4.1.9 with `@strapi/plugin-i18n": "4.1.9"`

**Comment by florianmrz** [comment]
### Pain point

**Issue:** 
Missing or very poor support for TypeScript.

**Context:**
(This might be biased, because I'm a big advocate of TS)
As others pointed out, the documentation in regards to backend customization is very poorly written. There's multiple ways to do things (e.g. query the database) and all options or variants are often not documented, while examples often don't reflect real-life requirements. I often find myself in the Strapi source code to understand how certain parts of the system work since the documentation can't answer my questions.

In my personal opinion, for a modern project this size, I can't understand why it was not written with TypeScript or type annotations in mind. It not only prevents bugs from happening for the Strapi dev team, but it also serves as documentation for the developer that wants to interact with the Strapi APIs and customize their server code or admin panel.

Coming from projects with excellent TypeScript support like NestJS or NextJS, this was a real pain for me after I started to actually customize the Strapi backend. Making matters worse, some of the existing types in the v4.2 beta are just wrongfully typed (this is being fixed right now!), requiring extra work to prevent unexpected Type errors.

My wish for the future would be to more aggressively move towards TypeScript support by for example rewriting parts of the Strapi core to TS and emitting the type declarations. This can be done bit by bit without hurting the existing codebase while simultaneously benefiting the developer that wants to customize Strapi - be it JS or TS developers, as both can benefit from the type IntelliSense and thus more easily understand how Strapi works.

**Which version are you on:**
v4

**Comment by MaksZhukov** [comment]
### Pain point

**Issue:** 
Composite primary key for content types

**Context:** 
I can't set id field as composite primary key.
I need to add addition uuid field to handle the case

**Which version are you on:**
v4

**Comment by hagenderouen** [comment]
### Pain point

**Issue:** 
No support for setting root domain

**Context:** 
Our organization hosts node applications using IIS with IISnode. We are not able to deploy Strapi due to our directory structure.  There should be a way to configure the root domain so that Strapi can be deployed on a nested directory structure. 

**Which version are you on:**
v3

**Comment by Pablo-Aldana** [comment]
### Pain Point

1 - The new data structure. Now all the core endpoints returns the information with the structure {id, data,attributes...} in my opinion this is a step backwards and more over if you try to use typed languages.

Before you can have your web/mobile/api consumer with an object Book: { title, ISBN} you call find and this would return an array of books, that is.  you parsed it fromJSON and end of the story

Now with v4 it returns now returns the data, as well as the attributes and to map all the attributes you have two options:
Re-write all your endpoints to return just plain books, or create a middleware or ugly constructor in your consumer that transform the shinny object into a book JSON to make it a real Book.

2 - Now to populate fields on a relationship i.e : Book: {title, isbn, author(relationship)} you have to make public the method find on author which is in my opinion a security breach. If a do a GET /books/ to get my books with their author I do not want to expose GET /authors/ so that anyone could potentially retrieve all the authors from my system and to have a data breach. it goes against software principles.

**Comment by jurjendevries** [comment]
Pain point:
A unique field for a slug is unusable with localization on. When I am using the same slug (which is sometimes the case when the word in another language is the same) in a localized item it won't let me. So I would like to see an exception setting for this, because duplicate slugs in the same language will give problems.

Context:
Content editing

Which version are you on:
v4.1.12 - Node version v14.19.3

**Comment by Webbist-dev** [comment]
### Pain point

**Issue:** 
Count not factoring in filters

**Context:** 
In Rest & GraphQL count returns are incorrect when results are filtered by more than 2 related entities. 
Pull request [here](https://github.com/strapi/strapi/pull/13480)
Related to [#7631](https://github.com/strapi/strapi/issues/7631), [#13136](https://github.com/strapi/strapi/issues/13136) and [#13443](https://github.com/strapi/strapi/issues/13443)
Anyone wanting to filter and paginate results must have come accross this.

Also folders in the media library as a quality of life. 

**Which version are you on:**
v4

**Comment by sapkra** [comment]
### Pain point

**Issue:** 
Insufficient TypeScript support

**Context:** 
We are using TypeScript for all our projects but the TypeScript support of strapi is very bad if even available. If there are types then they are not even correct and do not match the JavaScript codebase. It would be great if you can convert the strapi codebase to TypeScript step by step. This would automatically solve most of these issues.

**Which version are you on:**
v4

**Comment by Charlesdoiron** [comment]
**Pain point**
Issue:
Dont'loose relational fields when using **fill from another locale**

**Context:**
I’m working on a restaurant menu app.
The menus are in several languages.
Each menu is a content type.
A menu has a one-to-many relationship with products.
Each product is in several languages.
A menu can have 300 products.
When I create the menu in English and then I fill from another locale in the menu in Spanish I lose the integrity of my 300 related products.
This means that the content managers are tearing themselves away from linking these 300 products once again.

Which version are you on:
v4

**Comment by lkho** [comment]
### Pain point

**Issue:** 
Schema migration is not production ready.

**Context:** 
In `develop` mode, whenever changes requires server restart to perform database sync, the sync is executed immediately at the next process restart. This workflow works fine for development purpose, but it is unusable for production deployments. When you have a clustered environment, the db sync logic is performed every time in every worker when they restarted, which makes deployment like this likely to cause conflicts.

**Solution:**
I noticed that there is a TODO in the `@strapi/database` codebase which should be solving this issue, but I just want to point out it here since I think it is a blocking issue for production deployment. I just can't use strapi for any production environments which requires a HA/clustered setup.

https://github.com/strapi/strapi/blob/8d642957729fd35606e47a3df171852b41d437f1/packages/core/database/lib/schema/index.js#L62-L91

**Which version are you on:**
v4.3.0-beta.1

**Comment by laurenskling** [comment]
### Components do not work from plugins 

**Issue:** 
I run most of my code from plugins. Nice and isolated. Models, services, controllers, everything can be run from plugins. But not components. Such a powerful tool, but incomplete in implementation. :(

**Which version are you on:**
v3, but I also don't see it in v4

**Comment by DrMint** [comment]
### Pain point

**Issue:** 
Allow relations to be ordered and maintain this order.

**Context:** 
I'm using a one-to-many relationship between `Contents` and `ContentGroup` (a ContentGroup can have one or multiple Contents, a content has only one ContentGroup). I want the contents in a group to be ordered. Right now, the only way to change the order is to remove the relationships and readd them in the desired order. Furthermore, if one content is updated, its position in the group is changed (it is placed at the end of its group). I've seen people suggesting using repeatable composants (which are properly orderable) with a relationship inside the composant. However, only one-sided relationships are posible within a component which isn't usable for my use case.

In short, allow the user to reorder items within a relationship (the same way it's possible to reorder repeatable composants). And maintain this order despite items being updated.

**Which version are you on:**
v4

**Comment by wiesson** [comment]
### Pain point

**Issue:** 
Extended integration of image CDN possibilities 

**Context:** 
With e.g. cloudinary, there is no need to upload multiple versions of an image, (for example as a thumbnail) because the CDN can resize it perfectly on the fly. 

<img width="686" alt="Screenshot 2022-06-30 at 07 11 02" src="https://user-images.githubusercontent.com/933231/176597570-693c3e2a-d961-4acf-8881-8f670015f765.png">

The admin view could request the image in the perfect size to display it properly. 

Example function:
```Typescript
const cloudinaryRoot = "https://res.cloudinary.com/my-project/image/upload/";

type ImageLoaderProps = {
  src: string;
  width: number;
  height?: number;
  quality?: number;
  crop?: boolean;
};

export function extendedCloudinaryImageLoader({
  src,
  width,
  height,
  quality,
  crop = true,
}: ImageLoaderProps) {
  const params = [
    "f_auto",
    "dpr_auto",
    "w_" + width,
    "q_" + (quality || "auto"),
  ];

  if (crop) {
    params.push("c_fill");
  } else {
    params.push("c_limit");
  }

  if (height) {
    params.push("h_" + height);
  }

  let paramsString = params.join(",") + "/";
  return `${cloudinaryRoot}${paramsString}${src}`;
}
```

**Which version are you on:**
v4

**Comment by polem** [comment]
Pain point
Issue:
Allow keeping extra indexes / extra tables / extra views or materialized views

Context:
Sometimes we need :
- use custom tables (because strapi schema doesn't fit all use cases), and we don't always need a CRUD admin panel on it, (for example to store poll results)
- add custom index to solves perf issues
- use a materialized view (postgres) to simplify custom db queries

**Comment by AndColla** [comment]
### Pain point

**Issue:** 
Allow to order elements in a relation.
**Context:** 
The biggest problem for me right now is that I cannot order elements in a relation. Example: I have a collection named "People" and I want to have a subset of People in a defined order.

**Which version are you on:**
v3 and v4

**Comment by robertherber** [comment]
### Pain Point
**Issue:**
IDs missing on many entities/levels in the GraphQL API

**Context:**
Most GraphQL clients use normalized caching (using __typename and id) more or less out of the box. Not having IDs on every level/entity makes this very cumbersome to handle.

**Which version are you on:**
v4

**Comment by djmtype** [comment]
> ### Pain point
> **Issue:** Need to be able to sort relations and media
> 
> **Context:** I NEED, to have the possibility to order my media or my relations (articles on a feed for example). This is the biggest pain point for me ATM
> 
> **Which version are you on:** both

This is the main reason I've turned to Directus

**Comment by amargoCactus** [comment]
### Pain point

**Issue:**
Unresponsive UI

**Context:**
Strapi is incredible but one of the things that has kept me from developing a project for a client with strapi is that the UI is not suitable for tablets or cell phones. When I asked why this is so, they told me that "it is not a priority since many have not asked for it" so today I do my bit to make this possible.

**Which version are you on:**
v4

**Comment by brandonbuttars** [comment]
### Pain point

**Issue:** 
Need native web socket support for a more reactive backend.

**Context:** 
I want to use Strapi as my backend on a few different projects but I need web sockets and I prefer not to alter the code much at all when using it.  I am using Svelte and want to be able to get real-time updates to push to my stores in real-time.  With most front end frameworks now being much more real-time, I'd really like the backend to feel just as reactive.  I've debated using Google FireBase but don't want the whole "Google" part of that.  I've been looking into Supabase which looks promising but seems overly complicated and has a much steeper learning curve and more complicated stack.  In my perfect world Strapi would have web socket support out of the box.

A single socket that transmits different topics and data through its updates would be ideal.  I don't think more endpoints is the way to go.  I'm mostly wanting it for when the data relative to the logged in user has updates come through it.  I've seen the Socket.IO plugin but there's a Stackoverflow answer that very thoroughly communicates the differences and illustrates why you might use native websockets over Socket.IO.

https://stackoverflow.com/questions/10112178/differences-between-socket-io-and-websockets

**Which version are you on:**
v3 or v4

**Comment by MRDRMUFN** [comment]
### Pain point

**Issue:**    
Media gallery should auto filter out disabled filetypes

**Context:**    
If I set a media field to only accept images then the gallery when opened should only show image files. 

**Which version are you on:**    
v3 and v4

**Comment by MRDRMUFN** [comment]
### Pain point

**Issue:**    
Entering time in the **datetime** field via a dropdown menu is painfully slow

**Context:**    
I don't mind the dropdown so much but it now defaults to 0:00 instead of 12:00 from v3 and we now cannot manually type in values to save time scrolling.

**Which version are you on:**    
v3 and v4

**Comment by ohbob** [comment]
Pain point:

On a new remote instance, you cannot create admin account with out any insane manipulations. It's just a hustle to create the instance locally, upload it and use only then,just because you cannot create the account.

**Comment by oriooctopus** [comment]
There are lots of unnecessary layers to data when using graphql in v4. Take a look at two version of the same query, one in V3 and V4:
<img width="291" alt="cd6a21d180a0110eff6280040cff2f350e6a8a32_2_581x500" src="https://user-images.githubusercontent.com/20331238/176957580-3107e222-70b0-4374-8de3-52f7863702b9.png">

**Comment by jibin2706** [comment]
> There are lots of unnecessary layers to data when using graphql in v4. Take a look at two version of the same query, one in V3 and V4:
> <img width="291" alt="cd6a21d180a0110eff6280040cff2f350e6a8a32_2_581x500" src="https://user-images.githubusercontent.com/20331238/176957580-3107e222-70b0-4374-8de3-52f7863702b9.png">
> This gets even worse in the code when you have to deal with nullability and typescript.

It's a major step back

**Comment by biswajitdasme** [comment]
Probably we don't have transactions in 4.2 yet! :(

**Comment by Isaacn123** [comment]
i think me my main pain is not being able to to customize the homepage remove some Components like the Grid component section showing  tutorials etc...
if you could help on that

**Comment by jbcrestot** [comment]
## Pain point
## #Issue:
There are no debug/error logs when the server doesn't successfully start if you have an issue in your graphQL customization.

### Context:
I use Strapi as a backend for a mobile app, but I consume it through graphQL, when I add customizations (new type, mutation middleware, etc) and the syntax is wrong; the server doesn't start and you have nothing to help you understand your mistakes.

### Which version are you on:
v4

**Comment by Ymx1ZQ** [comment]
#Issue:
Missing fundamental feature to sort elements in a relation.

#Context:
In order to sort elements in a relation, you have to make crazy workarounds leading to a very bad UX. 
Example: I want to sort a subset of elements (say team members) out of a collection (e.g. a collection of people)

#Which version are you on:
v3 and v4

**Comment by Aurelsicoko** [comment]
Thank you, thank you, thank you for all this fantastic constructive feedback, it helps a lot! Please be sure we are committed to making Strapi a better CMS for all of you 🙏 

Please find my answer below where I tried to give as many explanations as possible and visibility for the coming months on the most important pain points highlighted in this thread.

**New documentation structure**
We recently reorganized our documentation team, and we will work closely with you in order to gather more feedback to focus on your needs. We want to bring the documentation to the next level. We are starting a redesign, and we will consider a new structure as well. Finally, we plan to simplify it by removing unnecessary sections and clearly differentiate the official vs community contributions/plugins/guides.

**Content management capabilities (scheduling, drafts, etc)**
We focused our efforts on the DX for the last two years and also the v4. We accumulated delay in the content editing experience regarding the competition. In Q3, we have a dedicated team that will work on a new draft system where you will be able to have a draft and a published version in parallel. This work will also set the foundations for having review workflows, scheduling, etc. 

**Communication with the community (migration guides, etc)**
We could have done way better in our communication to explain why the migration guides weren't ready and why it took so much time to finally have them. Ditto for the migration scripts. I'd like to publicly apologize for our lack of transparency and slowness on this very important topic. We learned a lot, and I'll personally be more involved in our monthly community calls to answer your questions and give the necessary explanations about our decisions such as releasing the v4 without the migration guides and scripts.

**Ordering the relations and improving their performance**
We are currently working on it, the performance will be improved thanks to a pagination system as we were currently loading all the relations. It was making the UI unresponsive and slow. The relations will be part of the main layout, the component will be brand new, as users you will have more space and again this initiative will set up the new foundations to enable relation reordering, but also, on-the-fly entry creation, better filtering & sorting.

**Improve UI performance (too slow)**
It's often linked to the previous topic, we recently improved the build time and removed it when you don't customize the administration panel. We would like to have more information, is there any specific part of the administration panel that is slow? Is it related to the response time of a request or the size of the administration panel package?

**Responsive administration panel**
We heard you, the new Design System started to take into consideration the tablets. There is still a long journey to make it responsive. There is also an alternative that would be to develop a mobile application. In any case, even for development purposes, our UI needs to be responsive at least for tablets. For mobile, it's going to be more challenging. For instance, the Cloud is currently developed to support small screens.

**API structure**
A lot of you complained about it when we introduced it. We know it's redundant. Currently, this format is ahead of its time. We designed it to support a true draft & publish system, publication workflows, scheduling, internationalization, environments and channels content distribution. These `attributes` and `data` properties will be extremely useful in the near future. Once we will have introduced the previously listed features, we could reconsider the structure. Until now, we are confident we made the right choice. The use of the API is useful but we hope we won't have to break it again when the new features will be available. However, there is still a missing piece in the puzzle: the SDK. We would like encourage our users to use it (once available) to make the interaction with the Strapi API easier.

**More customization in terms of fields, menu, and overall UI**
Again, we started developing six years ago and in the v3 we opened a lot of our customization capabilities. To be honest, it was a mistake. That's why we wanted to not reproduce the same mistake in the v4. In order to do that, we introduced an abstraction layer called the Plugin API. It allows you to customize Strapi without updating the core files while keeping your customization working even when you upgrade your project. Customization is great but also a dangerous path. It often makes migrations harder and upgrades difficult. Our mission is to empower you to manage your content, and we want to make sure you will be able to enjoy the new features. We will slowly open the customization areas in the UI, however, we don't want to go too fast as we didn't introduce the new & main content management features such as multiple drafts, scheduling, reviews workflows, etc. We don't know exactly (we have ideas) which part of the UI will be reserved for those.

**Better validations at the field level**
To be transparent, it isn't our current priority. We are aligned that it's necessary, the custom fields that we plan to introduce in beta in September might do the trick for a moment. However, it's something we would have to consider, certainly next year 👍 

**Better DZ management with a lot of components + nested DZ**
A few weeks ago, we removed the limitation to not having more than 2 levels of nesting within the components. Even, if often a better content structure could be the solution, there are some use cases when there is no other choice. Regarding the DZ, having them nested could create performance issues. We would love to hear about the use cases behind that need.

For the Dynamic Zones with a lot of components, we feel you. We experience the same on our own website (strapi.io). I don't want to overpromise things here, it's on our radar. Any community suggestion or pull request is welcomed 😊 

**Folders in the ML**
The folders are currently in beta, there will be available for everyone on July 27th

**Data/schema migration**
A dedicated team is working on the schema migration this quarter. They will also provide a new version for the API Token to get rid of the Users & Permissions plugin to access the API in a secure way with custom permissions.

**Increase compatibility trust for plugins**
This year, we released the marketplace which is also available in-app. We even organized a Plugin Week to almost reach 100 plugins listed on it. Our plugin ecosystem is much bigger than that with more than 2,000 plugins detected. In Q3, we plan to have filters, categories, and badges to make the marketplace easily browsable and identified what are the recommended plugins by the Core team. In addition to that, we will display the number of downloads and certainly the GitHub stars. We hope it will increase your confidence as users to install them in your project.

**Advanced media management (multiple formats, dynamic filter with the API, etc)**
A few of you complained about it, again I don't want to overpromise, it's on our radar, unfortunately, we don't have any initiative towards an improvement of the media management this year, except for the media folders. Any community contributions are welcomed though 😊

**Comment by derrickmehaffy** [comment]
Please do feel free to continue adding pain points until the cut off date and of course additional votes on existing pain points is awesome too!

**Comment by spenya** [comment]
I would like to be able to set webhooks for specific models ... I can filter it on my application, but I just need to catch updates on products, and I'm managing a bunch of request for other content types I really don't care about ...

**Comment by sapkra** [comment]
### Pain point

**Issue:** 
I18N Plugin doesn't work within other plugins

**Context:** 
We are structuring our codebase in plugins but if we want to define a content-type in a plugin which should be localized an error is thrown because the i18n plugin doesn't support plugins yet. That's a big painpoint because we now need to have related code in separate folders/packages in our mono-repo which results in less readable code.

**Which version are you on:**
v4

**Comment by vinzenzweber** [comment]
### Pain point

**Issue:** 
The id of translated content is not unique.

**Context:** 
 User generated content, that related to editor created content will therefore be connected to a specific locale only. 
 Eg: An editor creates a recipe in English and German. The id's for the recipe locales will be different, although it is the same content. If a user now likes a recipe in English, I would link the user id to the recipe id. Once the user switches locale, the recipe would no longer show up in his liked recipes.

**Which version are you on:**
v3

**Comment by derrickmehaffy** [comment]
Thank you everyone for the amazing responses to this year's pain point issue.

I'll go ahead and close/lock this so that the team can start looking through and categorizing the feedback and prepare a retrospective.

Above is a response from our CPO but certainly we would like everyone on our team to read, review, and process the feedback that selectively targets their teams.

We look forward to your feedback again!
